chips/virtio/src/devices/virtio_input.rs

```diff
@@ -0,0 +1,174 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2025.
+
+//! Support for the VirtIO Input Device
+//!
+//! <https://docs.oasis-open.org/virtio/virtio/v1.2/csd01/virtio-v1.2-csd01.html#x1-3850008>
+//!
+//! This implementation assumes the input device is a keyboard that follows the
+//! [event code](https://www.kernel.org/doc/Documentation/input/event-codes.txt)
+//! format that Linux uses.
+
+use kernel::utilities::cells::OptionalCell;
+
+use crate::devices::{VirtIODeviceDriver, VirtIODeviceType};
+use crate::queues::split_queue::{SplitVirtqueue, SplitVirtqueueClient, VirtqueueBuffer};
+
+/// Event: separate events.
+const EV_SYN: u16 = 0;
+/// Event: state change of keyboard.
+const EV_KEY: u16 = 1;
+
+/// VirtIO for input devices (e.g., keyboards).
+pub struct VirtIOInput<'a> {
+    /// Queue of events from the device (e.g., keyboard).
+    eventq: &'a SplitVirtqueue<'static, 'static, 3>,
+    /// Queue of status updates from this driver (e.g., turn on LED).
+    statusq: &'a SplitVirtqueue<'static, 'static, 1>,
+    /// Buffer to hold status updates.
+    status_buffer: OptionalCell<&'static mut [u8]>,
+    /// Store keys sent across multiple events.
+    keys: [OptionalCell<(u16, bool)>; 2],
```

Comment:
What led this to be length 2? I think you decided to only bother with the first two changed keys, even if there are more than that. That's probably reasonable, but it would be nice to document that.

Reply:
A google search suggests that 2 and 6 are common choices: https://en.wikipedia.org/wiki/Key_rollover